### PR TITLE
Datagen API polish

### PR DESCRIPTION
**Pull request by robertbastian**
#3564

### CODE REVIEW

**Comment by Manishearth** [comment]
(happy to review if needed but at this point there's a lot of stuff going on in datagen and I've only kept up with the high level things)

File: provider/datagen/src/bin/datagen/args.rs [hunk 1]
```diff
@@ -351,18 +345,18 @@ impl Cli {
                 ),
             }
         } else if let Some(key_file_path) = &self.key_file {
-            log::warn!("The --key-file argument is deprecated. Use --options with a JSON file.");
+            log::warn!("The --key-file argument is deprecated. Use a config.json.");
             #[allow(deprecated)]
             config::KeyInclude::Explicit(
-                icu_datagen::keys_from_file(key_file_path)
+                icu_datagen::keys_from_bin(key_file_path)
```
Comment (sffc) [inline provider/datagen/src/bin/datagen/args.rs:351]: ?

File: provider/datagen/src/lib.rs [hunk 2]
```diff
@@ -300,6 +373,11 @@ pub fn keys<S: AsRef<str>>(strings: &[S]) -> Vec<DataKey> {
 /// # Ok(())
 /// # }
 /// ```
+#[deprecated(
+    since = "1.3.0",
+    note = "use `all_keys` with the required cargo features"
```
Comment (sffc) [inline provider/datagen/src/lib.rs:378]: wrong deprecation message?

File: provider/datagen/src/bin/datagen/args.rs [hunk 3]
```diff
@@ -71,12 +71,10 @@ impl CollationTable {
...
 #[command(author = "The ICU4X Project Developers", version = option_env!("CARGO_PKG_VERSION"))]
 #[command(about = format!("Learn more at: https://docs.rs/icu_datagen/{}", option_env!("CARGO_PKG_VERSION").unwrap_or("")), long_about = None)]
-#[command(group(
-            ArgGroup::new("key_mode")
-                .required(true)
-                .args(["keys", "key_file", "keys_for_bin", "all_keys", "config"]),
-        ))]
 pub struct Cli {
+    #[arg(help = "Load a JSON config. All options other than --verbose are ignored.")]
+    config: Option<PathBuf>,
```
Comment (sffc) [inline provider/datagen/src/bin/datagen/args.rs:76]: Observation: you made this a positional argument

File: provider/datagen/src/lib.rs [hunk 4]
```diff
@@ -129,9 +202,9 @@ pub(crate) mod rayon_prelude {
 ///
 /// | Mode | Runtime Fallback | Data Size |
 /// |---|---|---|
-/// | [`Runtime`] | Yes, Automatic | Smallest |
-/// | [`RuntimeManual`] | Yes, Manual | Smallest |
-/// | [`Preresolved`] | No | Small |
+/// | [`Runtime`] | Required, Automatic | Smallest |
+/// | [`RuntimeManual`] | Required, Manual | Smallest |
+/// | [`Preresolved`] | Not required | Small |
```
Comment (sffc) [inline provider/datagen/src/lib.rs:207]: The problem with runtime fallback on preresolved is that we don't guarantee that things like `en-001` are present. You really shouldn't be using preresolved if you plan on runtime fallback.

File: provider/datagen/src/lib.rs [hunk 5]
```diff
@@ -308,9 +386,12 @@ pub fn keys_from_file<P: AsRef<Path>>(path: P) -> std::io::Result<Vec<DataKey>>
...
 
-/// Parses a compiled binary and returns a list of [`DataKey`]s used by it.
+/// Parses a compiled binary and returns a list of [`DataKey`]s that it uses *at runtime*.
 ///
-/// Unknown key names are ignored.
+/// This function is intended to be used for binaries that use `AnyProvider` or `BufferProvider`,
+/// for which the compiler cannot remove unused data. Using this function on a binary that only
+/// uses compiled data (through the `compiled_data` feature or manual baked data) will not return
+/// any keys, as the keys only exist in the type system.
 //  Supports the hello world key
```
Comment (sffc) [inline provider/datagen/src/lib.rs:395]: Nit: Clean up this hello world aside while you're in here

File: provider/datagen/src/provider.rs [hunk 6]
```diff
@@ -12,16 +12,16 @@ use std::fmt::Debug;
...
-/// If a specific data source has not been set, `DataProvider::load` will
-/// error ([`is_missing_cldr_error`](crate::is_missing_cldr_error) /
-/// [`is_missing_icuexport_error`](crate::is_missing_icuexport_error)) /
-/// [`is_missing_segmenter_lstm_error`](crate::is_missing_segmenter_lstm_error))
-/// if the data is required for that key.
+/// If a required data source has not been set, `DataProvider::load` will
+/// fail with the appropriate error:
+/// * [`is_missing_cldr_error`](Self::is_missing_cldr_error)
+/// * [`is_missing_icuexport_error`](Self::is_missing_icuexport_error)
+/// * [`is_missing_segmenter_lstm_error`](Self::is_missing_segmenter_lstm_error)
```
Comment (sffc) [inline provider/datagen/src/provider.rs:24]: Why did you move these onto `DatagenProvider`? I think they were good as free functions in the error module

**Review by sffc — COMMENTED** [review]
Seems good; some minor comments/questions

File: provider/datagen/src/lib.rs (hunk 4 above)
Comment (robertbastian) [reply provider/datagen/src/lib.rs:207]: do you have a suggestion? I'd rather say required for the first two

File: provider/datagen/src/lib.rs (hunk 5 above)
Comment (robertbastian) [reply provider/datagen/src/lib.rs:395]: it's not a doc comment

File: provider/datagen/src/lib.rs (hunk 4 above)
Comment (sffc) [reply provider/datagen/src/lib.rs:207]: Now about "N/A" or "Not applicable"

File: provider/datagen/src/provider.rs (hunk 6 above)
Comment (robertbastian) [reply provider/datagen/src/provider.rs:24]: Free functions are messy. These errors are a property of the DatagenProvider, not of any other part of datagen.

File: provider/datagen/src/lib.rs (hunk 5 above)
Comment (sffc) [reply provider/datagen/src/lib.rs:395]: If it's not a doc comment it should ideally not be in the middle of a doc comment

File: provider/datagen/src/bin/datagen/args.rs (hunk 1 above)
Comment (robertbastian) [reply provider/datagen/src/bin/datagen/args.rs:351]: ? indeed

File: provider/datagen/src/lib.rs (hunk 4 above)
Comment (robertbastian) [reply provider/datagen/src/lib.rs:207]: Not supported? If it's actually a problem we should discourage people from doing it

Comment (sffc) [reply provider/datagen/src/lib.rs:207]: Ok

**Review by sffc — APPROVED** [review]
lgtm modulo the preresolved docs
